llvm/lib/Target/RISCV/RISCVInstrInfoCOREV.td

```diff
@@ -431,57 +429,165 @@ def macu   : PatFrag<(ops node:$rd, node:$rs1, node:$rs2),
+            (CV_MACUN GPR:$rd, GPR:$rs1, GPR:$rs2, uimm5:$imm5)>;
+  def : Pat<(srl (machhu GPR:$rd, GPR:$rs1, GPR:$rs2), uimm5:$imm5),
+            (CV_MACHHUN GPR:$rd, GPR:$rs1, GPR:$rs2, uimm5:$imm5)>;
+  def : Pat<(ushiftRound (macu GPR:$rd, GPR:$rs1, GPR:$rs2), uimm5:$imm5),
+            (CV_MACURN GPR:$rd, GPR:$rs1, GPR:$rs2, uimm5:$imm5)>;
+  def : Pat<(ushiftRound (machhu GPR:$rd, GPR:$rs1, GPR:$rs2), uimm5:$imm5),
+            (CV_MACHHURN GPR:$rd, GPR:$rs1, GPR:$rs2, uimm5:$imm5)>;
+
+}
+
+//===----------------------------------------------------------------------===//
+// Patterns for general ALU operations
+//===----------------------------------------------------------------------===//
+
+let Predicates = [HasExtXCoreVAlu] in {
+
+  def : Pat<(abs GPR:$rs1), (CV_ABS GPR:$rs1)>;
+  def : PatGprGpr<setle, CV_SLET>;
+  def : PatGprGpr<setule, CV_SLETU>;
+  def : PatGprGpr<smin, CV_MIN>;
+  def : PatGprGpr<umin, CV_MINU>;
+  def : PatGprGpr<smax, CV_MAX>;
+  def : PatGprGpr<umax, CV_MAXU>;
+
+  def : Pat<(sext_inreg GPR:$rs1, i16), (CV_EXTHS GPR:$rs1)>;
+  def : Pat<(sext_inreg GPR:$rs1, i8), (CV_EXTBS GPR:$rs1)>;
+  def : Pat<(and GPR:$rs1, 0xffff), (CV_EXTHZ GPR:$rs1)>;
+  def : Pat<(and GPR:$rs1, 0xff), (CV_EXTBZ GPR:$rs1)>;
+
+  def : Pat<(clip powerOf2Minus1:$upperBound, GPR:$rs1),
+            (CV_CLIP GPR:$rs1, (trailing1sPlus1 imm:$upperBound))>;
+  def : Pat<(between (not GPR:$rs2), GPR:$rs2, GPR:$rs1),
+            (CV_CLIPR GPR:$rs1, GPR:$rs2)>;
+  def : Pat<(betweenu powerOf2Minus1:$upperBound, GPR:$rs1),
+            (CV_CLIPU GPR:$rs1, (trailing1sPlus1 imm:$upperBound))>;
+  def : Pat<(betweenu GPR:$rs2, GPR:$rs1),
+            (CV_CLIPUR GPR:$rs1, GPR:$rs2)>;
+
+  def : Pat<(sra (add GPR:$rs1, GPR:$rs2), uimm5:$imm5),
+            (CV_ADDN GPR:$rs1, GPR:$rs2, uimm5:$imm5)>;
+  def : Pat<(srl (add GPR:$rs1, GPR:$rs2), uimm5:$imm5),
+            (CV_ADDUN GPR:$rs1, GPR:$rs2, uimm5:$imm5)>;
+  def : Pat<(shiftRound (add GPR:$rs1, GPR:$rs2), uimm5:$imm5),
+            (CV_ADDRN GPR:$rs1, GPR:$rs2, uimm5:$imm5)>;
+  def : Pat<(ushiftRound (add GPR:$rs1, GPR:$rs2), uimm5:$imm5),
+            (CV_ADDURN GPR:$rs1, GPR:$rs2, uimm5:$imm5)>;
+
+  def : Pat<(sra (sub GPR:$rs1, GPR:$rs2), uimm5:$imm5),
+            (CV_SUBN GPR:$rs1, GPR:$rs2, uimm5:$imm5)>;
+  def : Pat<(srl (sub GPR:$rs1, GPR:$rs2), uimm5:$imm5),
+            (CV_SUBUN GPR:$rs1, GPR:$rs2, uimm5:$imm5)>;
+  def : Pat<(shiftRound (sub GPR:$rs1, GPR:$rs2), uimm5:$imm5),
+            (CV_SUBRN GPR:$rs1, GPR:$rs2, uimm5:$imm5)>;
+  def : Pat<(ushiftRound (sub GPR:$rs1, GPR:$rs2), uimm5:$imm5),
+            (CV_SUBURN GPR:$rs1, GPR:$rs2, uimm5:$imm5)>;
+
+  def : Pat<(sra (add GPR:$rd, GPR:$rs1), GPR:$rs2),
+            (CV_ADDNR GPR:$rd, GPR:$rs1, GPR:$rs2)>;
+  def : Pat<(srl (add GPR:$rd, GPR:$rs1), GPR:$rs2),
+            (CV_ADDUNR GPR:$rd, GPR:$rs1, GPR:$rs2)>;
+  def : Pat<(sra (add (add GPR:$rd, GPR:$rs1), (roundBit GPR:$rs2)), GPR:$rs2),
```

Comment:
Nothing prevents the `$rs2` from being `>=32`. Should round bit include a `and GPR:$rs2,  0x1f`?

Reply:
IIUC the shift operation is [undefined](https://llvm.org/docs/LangRef.html#id148) on shift amounts `>=32`, so we don't need to take care of that case.

Reply:
TIL. Yeah, makes sense, so nothing to do here.